File: Bloker63/Cards.py
import random


# -----------------------------------------------
#       ~                              ~
#               Suits and Ranks
#       ~                              ~
# -----------------------------------------------
Suits = [
    "WILD",
    "Spades",
    "Hearts",
    "Clubs",
    "Diamonds"
]

Ranks = [
    "WILD",
    "Joker",
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
    "Jack",
    "Queen",
    "King",
    "Ace"
]


# -----------------------------------------------
#       ~                              ~
#                   Classes
#       ~                              ~
# -----------------------------------------------
class Card:
    def __init__(self, suit: int = 0, rank: int = 0):
        self.suit: int = suit
        self.rank: int = rank

    def __lt__(self, other):
        return self.suit * 15 + self.rank < other.suit * 15 + other.rank

    def __repr__(self):
        if self.rank == self.suit == 0:
            return "Wild"

        if self.rank == 0:
            return f"Wild {Suits[self.suit]}".removesuffix("s")

        if self.suit == 0:
            return f"Wild {Ranks[self.rank]}"

        return f"{Ranks[self.rank]} of {Suits[self.suit]}"


class Deck:
    def __init__(self):
        self.Size = 0
        self.Cards: list[Card] = []
        self._drawn: list[Card] = []

    def add(self, card: Card, addToDiscardPile=False) -> None:
        if addToDiscardPile:
            self._drawn.append(card)
        else:
            self.Cards.append(card)

        self.Size += 1

    def clear(self):
        self.Cards.clear()
        self._drawn.clear()
        self.Size = 0

    def combine(self, other) -> None:
        """Put all cards from 'other' Deck into this deck"""

        self.Size += other.Size
        self.Cards.extend(other.Cards)
        self.Cards.extend(other._drawn)

        other.clear()

    def shuffle(self) -> None:
        random.shuffle(self.Cards)

    def sort(self, reverse=False) -> None:
        self.Cards.sort(reverse=reverse)

    def draw(self) -> Card:
        card = self.Cards.pop(random.randint(0, len(self.Cards)-1))
        self._drawn.append(card)
        return card

    def reset(self) -> None:
        self.Cards.extend(self._drawn)
        self._drawn.clear()

    def __repr__(self):
        if len(self.Cards) > 5:
            return f"Deck ({len(self.Cards)}/{self.Size}):\n - " + "\n - ".join([repr(c) for c in self.Cards[:5]])

        return f"Deck ({len(self.Cards)}/{self.Size}):\n - " + "\n - ".join([repr(c) for c in self.Cards])


# -----------------------------------------------
#       ~                              ~
#                 Functions
#       ~                              ~
# -----------------------------------------------
def CreateStandardDeck() -> Deck:
    d = Deck()

    for suit in range(1, 5):
        for rank in range(2, 15):
            c = Card(suit, rank)
            d.add(c)

    return d


__all__ = ["Ranks", "Suits", "Card", "Deck", "CreateStandardDeck"]

File: Baloker Online/src/modules/cards.c
#include "SDL3/SDL_stdinc.h"
#include <stdlib.h>
#include <stdio.h>
#include "cards.h"

#include "../baloker.h"

const char *SuitNames[numSuits] = {
    "Spades",
    "Hearts",
    "Clubs",
    "Diamonds"
};

const char *RankNames[numRanks] = {
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
    "Jack",
    "King",
    "Queen",
    "Ace"
};

Deck *CreateDeck()
{
    Deck *deck = malloc(sizeof(Deck));  // TODO: Check for malloc failure
    deck->cardCount = 0;
    return deck;
}

PlayerHand *CreateHand()
{
    PlayerHand *Hand = malloc(sizeof(PlayerHand));
    Hand->handCount = 0;
    Hand->riverCount = 0;

    return Hand;
}


Deck *CreateStandardDeck()
{
    Deck *deck = CreateDeck();

    for (int suit = 0; suit < numSuits; suit++)
    {
        for (int rank = 0; rank < numRanks; rank++)
        {
            int idx = suit * numRanks + rank;

            Card card = {0};
            card.rank = rank;
            card.suit = suit;

            card.target_scale = 1.0;
            card.default_scale = 1.0;
            card.scale = 1.0;

            card.tx = DeckPosX;
            card.ty = DeckPosY;

            card.x = card.tx;
            card.y = card.ty;

            card.seed = SDL_rand(99999);
            card.flipped = 1;
            card.highlighted = 0;
            card.ID = idx;

            deck->Cards[idx] = card;
            deck->cardCount += 1;
        }
    }

    return deck;
}

void printCard(Card card)
{
    printf("%s of %s\n", RankNames[card.rank], SuitNames[card.suit]);
}

void shuffleDeck(Deck* deck)
{
    for (int i = deck->cardCount - 1; i > 0; i--)
    {
        int j = SDL_rand(1000) % (i + 1);

        // swap cards[i] and cards[j]
        Card temp = deck->Cards[i];
        deck->Cards[i] = deck->Cards[j];
        deck->Cards[j] = temp;
    }
}

Card drawFromDeck(Deck* deck)
{
    Card card = deck->Cards[deck->cardCount-1];
    deck->cardCount -= 1;
    return card;
}

void addToDeck(Deck* deck, Card card)
{
    if (deck->cardCount < maxDeckSize)
    {
        deck->Cards[deck->cardCount] = card;
        deck->cardCount += 1;
    }
}

void addToHand(PlayerHand* player, Card card)
{
    if (player->handCount < maxHandSize)
    {
        player->Hand[player->handCount] = card;
        player->handCount += 1;
    }
}

void addToRiver(PlayerHand* player, Card card)
{
    if (player->riverCount < maxRiverSize)
    {
        player->River[player->riverCount] = card;
        player->riverCount += 1;
    }
}

void Deck_Destroy(Deck* deck)
{
    free(deck);
}

void Hand_Destroy(PlayerHand* hand)
{
    free(hand);
}

